Feat/Scrum75 카카오로그인 인증 코어 로직 구현, Feat/Scrum78 카카오로그인 인터페이스 작성

inwoodev opened the pull request:
## PR 타입
어떤 변경에 대한 PR인가요?

<!-- 아래 체크리스트 중 해당 되는 부분에 체크해주세요 "x". -->

- [ ] 버그 수정
- [x] Feature / 신기능
- [ ] 코드 스타일 업데이트 (formatting, local variables)
- [x] 리팩토링 (no functional changes, no api changes)
- [x] 빌드 관련 변경사항
- [ ] CI 관련된 변경사항
- [ ] Documentation content changes
- [ ] 구조 변경
- [ ] Other... 추가 설명 요망:


## 배경
[Feelin의 사용자 로그인API문서](https://www.notion.so/cdf6e46a61b149dbbc286a97892d3629?pvs=4)에 대응하여 카카오 로그인 & 서버 인증 로직을 구현해야 합니다.

## 목표
- 카카오 로그인에 필요한 KakaoSDK설정
- Tuist를 활용하여 QA, PROD, DEV 배포환경 세팅
- 다음과 같은 api flow 구성
  1. 카카오 로그인
  2. 카카오인증을 통해 전달받은 accessToken을 우리 서버로 전송
  3. 서버로부터 받은 accessToken & RefreshToken을 내부 키체인에 저장.



## 결과 (Optional)
- 배포 타겟 별 BundleIdentifier, kakaoNativeKey 등등 관련 Configuration이 달라질 수 있도록 세팅
- 카카오 로그인시 카카오앱으로 전환될 수 있는 URL Scheme설정 자동화
- KakaoOAuthService & KakaoLoginUsecase 구현 & 유닛테스트 수행

## Trouble or Trouble Shooting
현 시점 Domain과 Network에서 bundle을 활용하여 xcconfig의 환경변수에 접근하여 서버 base url 및 accessTokenKey를 가져올 수 있도록 세팅을 하려 했으나, 해당 작업이 뜻대로 되지 않았습니다. 이 문제는 차차 해결해야 합니다.

inwoodev committed:
refactor<DependencyInjection> DependencyInjection을 별도 모듈로 분리

- 불필요하게 DependencyInjection 모듈을 의존하는 부분 제거
- Core쪽 코드를 사용할 수 있도록 dependency 추가

inwoodev committed:
add<KakaoOpenSDK>: 카카오 소셜 로그인에 필요한 KakaoSDKCommon, KakaoSDKAuth, KakaoSDKUser 추가

- 외부 package들을 AppTarget에서 ThirdPartyLib로 옮김
- 카카오 로그인에 필요한 info.plist 내용 추가(LSApplicationQueriesSchemes)
- package.resolved 업데이트

inwoodev committed:
feat : Tuist를 활용한 QA, PROD, DEV 배포환경 세팅

- Scheme를 설정하는 ProjectDeescriptoinHelper Class 구현 및 적용
- PROD, DEV, QA, Sahred xcconfig 파일 추가
- Tuist의 Project 모듈 생성시 xcconfig를 포함한 config 세팅 추가
- 하나의 Feelin App Target을 Feelin-DEV, Feelin-QA, Feelin-PROD로 세분화 작업 및 Scheme 적용

inwoodev committed:
feat: 카카오 로그인을 위한 URLScheme, URLType 설정, 그리고 네이티브 앱 키 설정

- Shared.xcconfig에 카카오네이티브앱키를 호출할 수 있도록 키 설정을 하였다. 네이티브앱 키는 민감한 데이터를 누락하였고 빌드시 kakao developers로부터 앱 키를 가져와서 주입하여야 한다.

inwoodev committed:
feat : 카카오 네이티브 키를 관리할 xcconfig를 분리하여 관리

- 본래는 Shared.xcconfig를 앱키 없이 올린 뒤 .gitignore설정을 하려 하였으나, 그렇게 되면 Shared자체에 추후 추가될 설정들 또한 다른 개발자들에게 공유가 되지 않을 것이기에 잘못된 방향이라고 생각함.
- KakaoSecretKey라는 xcconfig를 생성하여 해당 설정을 통해서만 네이티브 앱키를 관리하는 방향으로 환경설정 수정
- KakaoSecretKey 파일을 올린 뒤 이후 .gitignore처리를 통해 네이티브 앱 키 유출을 방지할 예정

inwoodev committed:
refactor: 카카오 developer에서 등록된 PROD, DEV, QA 앱의 각기 다른 앱키에 대응할 수 있게 수정

- 각 배포 타겟 (Feelin, Feelin-DEV, Feelin-QA)마다 서로 다른 네이티브 앱 키를 사용해야 합니다. 따라서 각 타겟의 urlType의 urlSchemes에 해당 키를 제공해야 함.
- Info.plist에 작성한 KAKAO_NATIVE_APP_KEY를 유지하고, 배포 타입에 따라 이 키의 값을 각기 다른 xcconfig 파일의 프로퍼티에서 읽어오도록 설정

inwoodev committed:
chore: KakaoSecretKeys.xcconfig 파일을 gitignore & untrack

- .gitIgnore에 경로 추가
- git rm --cached 기능을 활용하여 해당 파일 추적 금지

inwoodev committed:
feat: Kakao로그인 초기화 세팅

- xcconfig에 저장된 카카오 네이티브 앱키를 사용하여 KakaoSDK초기화 진행
- 카카오 로그인 진행 후 처리 진행하기 위한 SceneDelegate의 `handleOpenURL()`처리
- Bundle에서 kakaoNativeAppKey를 불러오는 extension 메서드 추가

inwoodev committed:
add<DomainOAuth> 모듈 생성 및 의존성 세팅

그외:
- 불필요한 dependency 정리

inwoodev committed:
fix: xcconfig 커스텀 설정 후 유닛테스트가 실행이 안되는 문제 해결

inwoodev committed:
fix: Domain쪽 dependency오류 해결하기 위한 tuist 추가 설정

- 모든 도메인 모율의 dependency에 KakaoSDK추가(없으면 에러발생 추후 수정 필요할 것)
- 타겟생성시 test target같은 경우 Configuration에서 ENABLE_TESTABILITY를 true로 설정하기 위해 devSetting 적용
- Shared.xcconfig에 배포타겟 별 임시 url 추가(추후 수정해야 함)
- Domain과 Core 모듈간의 의존성 재설정
- AppDelegate에서 현시점 불필요한 dependency 제거

inwoodev committed:
refactor, feat: 로그인 API추가 및 NetworkProvider 개선

- api문서에 정의된 대로 body parameter에 auth_provider를 requestbody에 주입하여 post 수행
- api문서에 정의된 대로 잘못된 응답시 받는 커스텀에러를 디코딩하여 NetworkError로 변환하는 기능 추가
- URLRequestInterceptor를 NetworkProvider에 추가하여 필요시 api콜 중 urlrequest를 가로채 토큰을 header에 주입할 수 있고
  이를 활용하여 추후 토큰이 만료될 경우 재발급 받은 토큰을 처리할 수 있도록 기능 추가

inwoodev committed:
test: NetworkProvider &  Interceptor 그리고 APIFailResponse 에 대한 유닛테스트 추가

-

inwoodev committed:
test: 누락된 Combine+XCTest 파일 추가

- Tuist를 활용하여 Util모듈에서 XCTest를 선택적으로 의존할 수 있게 추가(Util에 XCTest를 import할 수 있게 하는 설정)
- combine을 확장하여 combine 사용시 테스트 길이를 간소화해주는 extension입니다.

inwoodev committed:
feat: KakaoOAuthService & KakaoLoginUsecase 추가

inwoodev committed:
test: KakaoOAuthService & JWTDecoder 유닛테스트 추가

inwoodev commented on Plugins/DependencyPlugin/ProjectDescriptionHelpers/Project+Environment.swift:
```diff
@@ -38,33 +38,45 @@ public extension Project {
+                kakaoNativeAppKey =  "${KAKAO_NATIVE_APP_KEY_QA}"
+            case .prod:
+                kakaoNativeAppKey =  "${KAKAO_NATIVE_APP_KEY_PROD}"
+            }
```
배포 타겟에 따라 다른 kakaoNativeAppKey를 지정합니다.

[카카오 개발](https://developers.kakao.com/) 페이지에서 내 애플리케이션에서 각 타겟의 bundleIdentifier에 맞는 kakaoNativeAppKey를 확인할 수 있습니다.

derrickkim0109 replied:
Apple 로그인 API Key도 여기에서 관리하면 될까요?

inwoodev commented on Projects/App/xcconfigs/KakaoSecretKeys.xcconfig:
```diff
@@ -9,4 +9,6 @@
-KAKAO_NATIVE_APP_KEY = 
+KAKAO_NATIVE_APP_KEY_PROD =
+KAKAO_NATIVE_APP_KEY_DEV =
+KAKAO_NATIVE_APP_KEY_QA =
```
이 변수의 value는 깃헙에 올라오면 안됩니다!

inwoodev committed:
move: JWT관련 모델들을 네트워크 인터페이스 모듈로 이관

추가될 interceptor에서도 JWT 파싱처리를 할 것이라 예상. 네트워크 모듈의 Token Interceptor & Service 모듈에서 모두 접근 가능하도록 파일 이동

inwoodev committed:
feat: NetworkSession & TokenInterceptor 기능 추가

inwoodev committed:
refactor: endpoint에 header추가 & 로그인 관련 로직 수정

- 카카오 로그인을 통해 전달받은 accessToken을 볍도의 interceptor없이 주입할 수 있도록 api 및 구조 수정

inwoodev committed:
refactor: 서버의 커스텀 에러메세지를 NetworkError로 변환하는 코드 추가

inwoodev committed:
tess: NetworkSession 그리고 분리된 Interceptor구조에 대응할 수 있게 테스트 &테스팅 기능 수정

inwoodev committed:
feat: 토큰 검증 및 재발급 api  초기 구현

inwoodev committed:
add, test: 변경된 토큰 만료 & 재발급 로직에맞게 관련 테스트코드 수정

inwoodev committed:
test: UserValidity관련 통합테스트를 간단하게 구현

inwoodev commented on Projects/Core/Network/Interface/Sources/NetworkSession.swift:
```diff
@@ -0,0 +1,41 @@
+
+public typealias DataTaskResult = (data: Data, response: URLResponse)
+
+open class NetworkSession {
+    public let urlSession: URLSession
+    public var requestInterceptor: URLRequestInterceptor?
+    
+    public init(
+        urlSession: URLSession = .shared,
+        requestInterceptor: URLRequestInterceptor?
+    ) {
+        self.urlSession = urlSession
+        self.requestInterceptor = requestInterceptor
+    }
+    
+    public func dataTaskPublisher(for request: URLRequest) -> AnyPublisher<DataTaskResult, Error> {
+        if let interceptor = requestInterceptor {
+            return interceptor.adapt(request)
+                .flatMap { [unowned self] adaptedRequest -> AnyPublisher<DataTaskResult, Error> in
+                    return urlSession.dataTaskPublisher(for: adaptedRequest)
+                        .mapError { $0 as Error }
+                        .eraseToAnyPublisher()
+                }
+                .eraseToAnyPublisher()
+            
+        } else {
+            return urlSession.dataTaskPublisher(for: request)
+                .mapError { $0 as Error }
+                .eraseToAnyPublisher()
+        }
+    }
+}
```
Alamofire를 참고해서 urlSession을 한 번 더 래핑하여 `dataTaskPublisher`를 실행할 수 있도록 구현했어요.
장점이라하면 위 처럼 requestInterceptor 유무에 따라서 유동적으로 대응할 수 있습니다~

inwoodev commented on Projects/Core/Network/Sources/Extensions/Publisher+.swift:
```diff
@@ -67,7 +67,16 @@ public extension Publisher where Output == DataTaskResult {
         return tryMap { output in
             if let apiFailResponse = try? decoder.decode(APIFailResponse.self, from: output.data) {
-                throw NetworkError.customServerError(apiFailResponse)
+                switch apiFailResponse.errorCode {
+                case "00401":
+                    throw NetworkError.clientError(.authorizationError)
+                    
+                case "00500":
+                    throw NetworkError.serverError(.internalServerError)
+                    
+                default:
+                    throw NetworkError.customServerError(apiFailResponse)
+                }
```
서버로부터 받은 json data 디코딩 전 서버에서 정의한 양식의 에러메세지 타입으로 먼저 변환을 시도 후 성공한다면 서버에서 내려주는 에러코드에 따라 저희가 기존에 사용하는 에러타입으로 던지도록 기능을 추가했습니다.

이렇게 하면 저희가 기존에 추가했던 `NetworkError`를 굳이 수정하지 않아도 되어도 대응할 수 있다는 장점이 있다고 생각이 들었습니다~

derrickkim0109 replied:
`00401`, `00500` 텍스트는 따로 관리가 필요하지 않을까요?

inwoodev commented on Projects/Core/Network/Sources/TokenInterceptor.swift:
```diff
@@ -0,0 +1,112 @@
+import Foundation
+
+extension TokenInterceptor: URLRequestInterceptor {
+    public func adapt(_ urlRequest: URLRequest) -> AnyPublisher<URLRequest, Error> {
+        var request = urlRequest
+        
+        do {
+            guard let accessToken: AccessToken = try tokenStorage.read(key: accessTokenKey) else {
+                return Fail(error: NetworkError.urlRequestError(.headerError(reason: "키체인에 저장되어있는 액세스 토큰이 없습니다.")))
+                    .eraseToAnyPublisher()
+            }
+            request.setValue(
+                "Bearer \(accessToken)",
+                forHTTPHeaderField: "Authorization"
+            )
+            
+        } catch let error as KeychainError {
+            return Fail(error: NetworkError.requestInterceptError("Token Interceptor에서 액세스 토큰을 불러오는 과정에서 에러가 발생했습니다. 에러: \(error.errorDescription ?? error.localizedDescription)"))
+                .eraseToAnyPublisher()
+        } catch {
+            return Fail(error: NetworkError.requestInterceptError("Token Interceptor에서 알 수 없는 에러가 발생했습니다."))
+                .eraseToAnyPublisher()
+        }
+        
+        return Just(request)
+            .setFailureType(to: Error.self)
+            .eraseToAnyPublisher()
+    }
```
interceptor를 NetworkSession생성시 주입한다면 기본적으로 내부에 있는 토큰을 불러와서 request의 헤더에 세팅하도록 로직을 짜봤습니다.

inwoodev commented on Projects/Core/Network/Sources/TokenInterceptor.swift:
```diff
@@ -0,0 +1,112 @@
+        dueTo error: NetworkError
+    ) -> AnyPublisher<RetryResult, Never> {
+        guard case .clientError(.authorizationError) = error else {
+            return Just(.doNotRetry)
+                .eraseToAnyPublisher()
+        }
+        
+        var request = request
+        
+        do {
+            guard let refreshToken: RefreshToken = try tokenStorage.read(key: refreshTokenKey) else {
+                return Just(.doNotRetryWithError(NetworkError.requestInterceptError("키체인에 저장되어있는 액세스 토큰이 없습니다.")))
+                    .eraseToAnyPublisher()
+            }
+            request.setValue(
+                "Bearer \(refreshToken.token)",
+                forHTTPHeaderField: "Authorization"
+            )
+            
+            let reissueEndpoint = FeelinAPI<ReissueTokenResponse>.reissueAccessToken(refreshToken: refreshToken.token)
+            
+            let reissueRequest = try reissueEndpoint.makeURLRequest()
+            
+            return reissueToken(request: reissueRequest, with: session)
+            
+        } catch let error as NetworkError {
+            return Just(.doNotRetryWithError(error))
+                .eraseToAnyPublisher()
+            
+        } catch let error as KeychainError {
+            return Just(.doNotRetryWithError(NetworkError.requestInterceptError("Token Interceptor에서 액세스 토큰을 불러오는 과정에서 에러가 발생했습니다. 에러: \(error.errorDescription ?? error.localizedDescription)")))
+                .eraseToAnyPublisher()
+        } catch let error as JWTError {
+            return Just(.doNotRetryWithError(NetworkError.requestInterceptError("Token Interceptor에서 jwt파싱 과정에서 에러가 발생했습니다. 에러:\(error.errorDescription ?? error.localizedDescription) ")))
+                .eraseToAnyPublisher()
+        } catch {
+            return Just(.doNotRetryWithError(NetworkError.unknownError(error.localizedDescription)))
+                .eraseToAnyPublisher()
+        }
+    }
```
Interceptor를 사용할 때 에러를 캐치할 때 불리는 메서드에요.
내부에 저장되어있는 refreshToken을 활용해서 accessToken을 재발급을 할 수 있도록 구현을 했습니다.

이게 살짝 애매하긴 한데 이 과정에서 LocalStorage쪽 모듈들을 필요로 하고 KeychainError, JWTError와 같이 기존 네트워크에러에 포함되지 않는 에러들을 내뱉을 가능성이 있어서 고민하다가 NetworkError에 .requestInterceptError케이스를 추가하여 해당 에러에 자세한 메세지를 적는 방향으로 구현해 봤습니다!

inwoodev commented on Projects/Core/Network/Sources/TokenInterceptor.swift:
```diff
@@ -0,0 +1,112 @@
+        }
+    }
+    
+    private func reissueToken(
+        request: URLRequest,
+        with session: URLSession
+    ) -> AnyPublisher<RetryResult, Never> {
+        
+        return session
+            .dataTaskPublisher(for: request)
+            .tryDecodeAPIFailResponse()
+            .validateStatusCode()
+            .validateJSONValue(to: ReissueTokenResponse.self)
+            .map(\.data)
+            .tryMap { [jwtDecoder] tokenResponse -> (AccessToken, RefreshToken) in
+                return (
+                    try jwtDecoder.decode(tokenResponse.accessToken, as: AccessToken.self),
+                    try jwtDecoder.decode(tokenResponse.refreshToken, as: RefreshToken.self)
+                )
+            }
+            .tryMap { [accessTokenKey, refreshTokenKey, tokenStorage] (accessToken, refreshToken) -> RetryResult in
+                try tokenStorage.save(token: accessToken, for: accessTokenKey)
+                try tokenStorage.save(token: refreshToken, for: refreshTokenKey)
+                
+                return .retry
+            }
+            .catch { error in
+                return Just(.doNotRetryWithError(error))
+            }
+            .eraseToAnyPublisher()
+    }
```
요 메서드는 별도의 객체로 빼도 되지만 일단 테스트함에 있어서 크게 문제가 되지 않는다 판단하여 일단 내부 private 메서드로 구현했습니다. retry과정에서 accessToken을 재발급하는 메서드입니다. 재발급 받은 토큰을 내부에 저장하며 .retry를 반환하게 하여 기존 실패했던 api를 한 번 더 시도할 수 있게됩니다.

inwoodev commented on Projects/Core/Network/Testing/Sources/CountableMockURLProtocol.swift:
```diff
@@ -0,0 +1,46 @@
+//
+
+import Foundation
+
+class CountableMockURLProtocol: URLProtocol {
+    static var requestHandlers: [((URLRequest) -> (HTTPURLResponse, Data?))] = []
+    private static var requestCount = 0
+
+    override class func canInit(with request: URLRequest) -> Bool {
+        return true
+    }
+
+    override class func canonicalRequest(for request: URLRequest) -> URLRequest {
+        return request
+    }
+
+    override func startLoading() {
+        guard CountableMockURLProtocol.requestCount < CountableMockURLProtocol.requestHandlers.count else {
+            client?.urlProtocol(self, didFailWithError: URLError(.badServerResponse))
+            return
+        }
+
+        let handler = CountableMockURLProtocol.requestHandlers[CountableMockURLProtocol.requestCount]
+        CountableMockURLProtocol.requestCount += 1
+        let (response, data) = handler(request)
+
+        client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
+        if let data = data {
+            client?.urlProtocol(self, didLoad: data)
+        }
+        client?.urlProtocolDidFinishLoading(self)
+    }
+
+    override func stopLoading() {}
+
+    static func reset() {
+        requestCount = 0
+        requestHandlers = []
+    }
+}
```
URLProtocol을 상속받아서 여러 개의 requestHandler를 지정할 수 있도록 만들어 봤습니다~

inwoodev commented on Projects/Domain/Tests/Sources/UserValidityIntegrationTests.swift:
```diff
@@ -0,0 +1,146 @@
+                    {
+                        "message" : "액세스 토큰이 재발급되었습니다.",
+                        "data": {
+                            "accessToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNjg1NjA4MDAwLCJleHAiOjE3MTcwNDUyMDB9.ZsTGkCZQizZDC8E4K9nmdsk7KZJz0LbG-C5n7A2KbBs",
+                            "refreshToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiI1Njc4OTAxMjM0IiwibmFtZSI6IkphbmUgRG9lIiwiaWF0IjoxNjg1NjA4MDAwLCJleHAiOjE3MTcwNDUyMDB9.pE12rWqpm0Oa1owwymenRzg1YwE-AidcTWw1BrH6Kx8"
+                        }
+                    }
+                """.data(using: .utf8)!
+                
+                return (mockResponse, mockData)
+            },
+            { requestWithReissueToken in
+                let mockResponse = HTTPURLResponse(
+                    url: requestWithReissueToken.url!,
+                    statusCode: 201,
+                    httpVersion: nil,
+                    headerFields: nil
+                )!
+                
+                let mockData =
+                """
+                    {
+                        "message" : "액세스 토큰이 유효합니다.",
+                        "data" : {
+                            "isValid" : true,
+                            "expirationDate" : "2024-06-01T12:00:00Z"
+                        }
+                    }
+                """.data(using: .utf8)!
+                
+                return (mockResponse, mockData)
+            }
+        ]
+        
+        // when
+        let result = try awaitPublisher(sut.isUserValid())
+        
+        // then
+        XCTAssertTrue(result)
+    }
```
통합 유닛테스트를 간단하게 작성해 봤어요. 제가 생각했을 때는 제일 복잡한 로직이 `토큰 유효성 검증 부분`이라 생각이 들어서 이 부분만 일단 테스트 진행했습니다.

`CountableMockURLProtocol`을 활용하여 
- 첫 번째 dataTask의 결과(401에러, 토큰만료) 
- 두 번째 dataTask결과(새로운 토큰발급), 
- 그리고 마지막 dataTask결과(액세스토큰 유효성체크)
이렇게 3가지 요청에 대한 의도된 데이터를 반환하도록 설정하였습니다.

이를 활용하여 UserValidityService의 `isUserValid()`메서드 테스트를 진행 해 봤습니다 :)

inwoodev commented on Projects/Shared/Util/Sources/Extensions/Combine+XCTest.swift:
```diff
@@ -0,0 +1,58 @@
+        let expectation = self.expectation(description: "Awaiting publisher")
+
+        let cancellable = publisher.sink(
+            receiveCompletion: { completion in
+                switch completion {
+                case .failure(let error):
+                    result = .failure(error)
+                case .finished:
+                    break
+                }
+
+                expectation.fulfill()
+            },
+            receiveValue: { value in
+                result = .success(value)
+            }
+        )
+        
+        waitForExpectations(timeout: timeout)
+        cancellable.cancel()
+
+        let unwrappedResult = try XCTUnwrap(
+            result,
+            "Awaited publisher did not produce any output",
+            file: file,
+            line: line
+        )
+
+        return try unwrappedResult.get()
+    }
+}
+
+extension Published.Publisher {
+    func collectNext(_ count: Int) -> AnyPublisher<[Output], Never> {
+        self.dropFirst()
+            .collect(count)
+            .first()
+            .eraseToAnyPublisher()
+    }
+}
```
요거 활용하시면 유닛테스트 시 컴바인 사용하실 때 코드길이가 많이 줄어드실거에요 ㅎㅎ

derrickkim0109 replied:
예시 조금만 알려주실 수 있을까요?ㅎ

inwoodev commented on Projects/Core/Network/Sources/Extensions/Publisher+.swift:
```diff
@@ -49,7 +49,10 @@ public extension Publisher where Output == DataTaskResult {
     func validateJSONValue<Output: Decodable>(to outputType: Output.Type) -> AnyPublisher<Output, NetworkError> {
         return tryMap {
-            try JSONDecoder().decode(outputType, from: $0.data)
+            // MARK: - 서버와 date 형식을 어떻게 주고 받을 지 정의 해야 함.
+            let decoder = JSONDecoder()
+            decoder.dateDecodingStrategy = .iso8601
+            return try decoder.decode(outputType, from: $0.data)
```
하 이 작은 부분 때문에 어제 밤에 얼마나 끙끙 앓았는지 모릅니다 ㅠㅠ 유닛테스트할 때 iso8601로 표현된 날짜를 사용했는데 위 설정을 안해서 계속 decodingError가 떴던 기억이...무튼 날짜를 수신할 때 어떤 전략을 취할 지 저희도 논의 해 보면 좋을 것 같아요~

derrickkim0109 replied:
좋습니다~!

derrickkim0109 commented on Plugins/DependencyPlugin/ProjectDescriptionHelpers/Configuration.swift:
```diff
@@ -0,0 +1,43 @@
+            return .debug(
+                name: type.configurationName,
+                settings: [
+                    "SWIFT_ACTIVE_COMPILATION_CONDITIONS": "$(inherited) MOCKING",
+                    "ENABLE_TESTABILITY": true
```
FlexLayoutSDK 때문에 각 settings에 base로 GCC 전처리기 설정이 필요한데 이 부분은 만나서 같이 할까요?

inwoodev replied:
일단 Apple 로그인 키도 이렇게 관리할려고 생각 중입니다. 물론 추후 리팩토링을 통해서 개선 해 볼 수 있을 듯 합니다

derrickkim0109 commented on Plugins/DependencyPlugin/ProjectDescriptionHelpers/Project+Environment.swift:
```diff
@@ -12,19 +12,76 @@ public extension Project {
+            ]
+        )
+        
+        public static func appInfoPlist(deploymentTarget: ProjectDeploymentTarget) -> InfoPlist {
+            var kakaoNativeAppKey: String = ""
+            var baseServerURL: String = ""
+            switch deploymentTarget {
+            case .dev:
+                kakaoNativeAppKey = "${KAKAO_NATIVE_APP_KEY_DEV}"
+                baseServerURL = "${BASE_SERVER_URL_DEV}"
+            case .qa:
+                kakaoNativeAppKey =  "${KAKAO_NATIVE_APP_KEY_QA}"
+                baseServerURL = "${BASE_SERVER_URL_QA}"
+            case .prod:
+                kakaoNativeAppKey =  "${KAKAO_NATIVE_APP_KEY_PROD}"
+                baseServerURL = "${BASE_SERVER_URL_PROD}"
+            }
```
되게 잘 하신 것 같아요~ 굿굿

derrickkim0109 commented on Projects/Core/Network/Interface/Sources/Helper/JWTDecoder.swift:
```diff
@@ -0,0 +1,53 @@
+            .replacingOccurrences(of: "-", with: "+")
+            .replacingOccurrences(of: "_", with: "/")
+        
+        let padLength = (4 - base64.count % 4) % 4
+        base64.append(contentsOf: repeatElement("=", count: padLength))
```
이 부분에 대해 설명해주실 수 있을까요?

inwoodev replied:
[블로그글](https://mtanriverdi.medium.com/how-to-decode-jwt-and-validate-the-signature-in-swift-97092bd654f7) 한 번 참고하시면 도움이 되실듯 합니다 :)

derrickkim0109 commented on Projects/Core/Network/Sources/NetworkProvider.swift:
```diff
@@ -23,17 +23,26 @@ public final class NetworkProvider: NetworkProviderProtocol {
 
-            return session
+            return networkSession
                 .dataTaskPublisher(for: urlRequest)
+                .tryDecodeAPIFailResponse()
                 .validateStatusCode()
+                .retryOnUnauthorized(
+                    session: networkSession,
+                    request: urlRequest
+                )
```
**P4**
validateStatusCode보다 tryDecodeAPIFailResponse를 먼저 체크하는 이유가 있을까요?

derrickkim0109 commented on Projects/Core/Network/Sources/TokenInterceptor.swift:
```diff
@@ -0,0 +1,112 @@
+                .eraseToAnyPublisher()
+            
+        } catch let error as KeychainError {
+            return Just(.doNotRetryWithError(NetworkError.requestInterceptError("Token Interceptor에서 액세스 토큰을 불러오는 과정에서 에러가 발생했습니다. 에러: \(error.errorDescription ?? error.localizedDescription)")))
```
**P4**
"Token Interceptor에서 액세스 토큰을 불러오는 과정에서 에러가 발생했습니다. 에러:"이 문구는 KeychainError에서 같이 관리할 수는 없을까요?
아래의 JWKError도 동일합니다.

derrickkim0109 commented on Projects/Core/Network/Sources/TokenInterceptor.swift:
```diff
@@ -0,0 +1,112 @@
+        request: URLRequest,
+        with session: URLSession
+    ) -> AnyPublisher<RetryResult, Never> {
+        
```
여기 개행이 있어요~!

derrickkim0109 commented on Projects/Core/Network/Testing/Sources/CountableMockURLProtocol.swift:
```diff
@@ -0,0 +1,46 @@
+    static var requestHandlers: [((URLRequest) -> (HTTPURLResponse, Data?))] = []
+    private static var requestCount = 0
+
+    override class func canInit(with request: URLRequest) -> Bool {
+        return true
+    }
+
+    override class func canonicalRequest(for request: URLRequest) -> URLRequest {
+        return request
+    }
```
**P5**

이 메서드들을 사용하는 이유에 대해 설명해주실 수 있을까요?

derrickkim0109 commented on Projects/Core/Network/Testing/Sources/TokenInterceptorTestDoubles.swift:
```diff
@@ -0,0 +1,64 @@
+    }
+    
+    public func adapt(_ urlRequest: URLRequest) -> AnyPublisher<URLRequest, Error> {
+        self.adaptMethodCalled = true
+        var request = urlRequest
+        
+        do {
+            guard let accessToken: AccessToken = try tokenStorage.read(key: dummyAccessTokenKey) else {
+                return Fail(error: NetworkError.urlRequestError(.headerError(reason: "키체인에 저장되어있는 액세스 토큰이 없습니다.")))
+                    .eraseToAnyPublisher()
+            }
+            request.setValue(
+                "Bearer \(accessToken.token)",
+                forHTTPHeaderField: "Authorization"
+            )
+            self.request = request
+            
```
adaptMethodCalled 이 프로퍼티는 정상적으로 adapt가 호출되었다는 걸 감지하는 용도로 사용하는 걸까요?
그렇다면 request 프로퍼티에 값이 다 들어와야 true인게 아닐까요?

derrickkim0109 commented on Projects/Core/Network/Tests/Sources/RequestInterceptTests.swift:
```diff
@@ -0,0 +1,141 @@
+    var fakeTokenStorage: FakeTokenStorage!
+    var tokenInterceptor: MockTokenInterceptor!
+    var urlSession: URLSession!
+    
+    
```
개행이 있어요~

derrickkim0109 commented on Projects/Core/Network/Tests/Sources/RequestInterceptTests.swift:
```diff
@@ -0,0 +1,141 @@
+        
+        let endpoint = FeelinAPI<UserValidityResponse>.checkUserValidity
+        
+        
```
여기도 개행이 있는데, 

**P5**
테스트 코드에도 다는건 조금 그럴까요..?

derrickkim0109 commented on Projects/Core/Project.swift:
```diff
@@ -14,9 +14,7 @@ let targets: [Target] = [
                 .core(implements: .Network),
-				.core(implements: .DependencyInjection),
-				.core(implements: .LocalStorage),
-                .shared(implements: .Util)
+				.core(implements: .LocalStorage)
 			]
```
정렬이 안된 것 같네요. 한번 해주실 수 있나요?

derrickkim0109 commented on Projects/Domain/OAuth/Project.swift:
```diff
@@ -0,0 +1,58 @@
+        factory: .init(
+            dependencies: [
+                .domain(interface: .OAuth),
+                .SPM.KakaoSDKCommon,
+                .SPM.KakaoSDKAuth,
+                .SPM.KakaoSDKUser,
+            ]
+        )
+    ),
+    .domain(
+        testing: .OAuth,
+        factory: .init(
+            dependencies: [
+                .domain(interface: .OAuth),
+                .SPM.KakaoSDKCommon,
+                .SPM.KakaoSDKAuth,
+                .SPM.KakaoSDKUser,
+            ]
+        )
+    ),
+    .domain(
+        tests: .OAuth,
+        factory: .init(
+            dependencies: [
+                .domain(testing: .OAuth),
+                .domain(implements: .OAuth),
+                .domain(interface: .OAuth),
+                .SPM.KakaoSDKCommon,
+                .SPM.KakaoSDKAuth,
+                .SPM.KakaoSDKUser,
```
카카오 SDK도 .domain(interface: .OAuth)로 SDK에 대한 의존성 부여가 안되나요?

derrickkim0109 replied:
core 모듈에서 .shared를 가지도록 해서 올렸는데 이 부분은 같이 조금 더 해볼까요?

derrickkim0109 commented on Projects/Domain/OAuth/Interface/Sources/KakaoUserAPIProtocol.swift:
```diff
@@ -0,0 +1,31 @@
+import KakaoSDKUser
+
+public protocol KakaoUserAPIProtocol {
+    func loginWithKakaoTalk(
+        launchMethod: KakaoLaunchMethod?,
+        channelPublicIds: [String]?,
+        serviceTerms: [String]?,
+        nonce: String?,
+        completion: @escaping (KakaoOAuthToken?, Error?) -> Void
+    )
+    
+    func loginWithKakaoAccount(
+        prompts : [KakaoAuthPrompt]?,
+        channelPublicIds: [String]?,
+        serviceTerms: [String]?,
+        nonce: String?,
+        completion: @escaping (KakaoOAuthToken?, Error?) -> Void
+    )
```
KakaoUserAPI에는 Kakao와 관련된 메서드만 있을 수 있다면

**P3**
Protocol 내의 메서드 네이밍은 Kakao가 붙지 않아도 되지 않을까요?

derrickkim0109 commented on Projects/Domain/OAuth/Sources/Service/UserValidityService.swift:
```diff
@@ -0,0 +1,24 @@
+import DomainOAuthInterface
+
+extension UserValidityService: UserValidityServiceInterface {
+    public func isUserValid() -> AnyPublisher<Bool, AuthError> {
```
여기도 위의 KakaoUserAPI에서의 의견과 같습니다.

derrickkim0109 commented on Projects/Domain/OAuth/Tests/Sources/KakaoOAuthServiceTests.swift:
```diff
@@ -0,0 +1,111 @@
+            ),
+            error: nil
+        )
+        
```
여기도 개행이 있습니다.

derrickkim0109 commented on Projects/Shared/ThirdPartyLib/Project.swift:
```diff
@@ -16,13 +16,22 @@ let targets: [Target] = [
                 .SPM.Kingfisher,
                 .SPM.PinLayout,
-                .SPM.FlexLayout
+				.SPM.FlexLayout,
+				.SPM.KakaoSDKAuth,
+				.SPM.KakaoSDKUser,
+				.SPM.KakaoSDKCommon
             ]
```
정렬이 안된 것 같아요~

inwoodev replied:
수정했습니다~

derrickkim0109 commented on Projects/Shared/Util/Sources/Extensions/Combine+XCTest.swift:
```diff
@@ -0,0 +1,58 @@
+import XCTest
+
+extension XCTestCase {
+    func awaitPublisher<T: Publisher>(
+        _ publisher: T,
+        timeout: TimeInterval = 10,
+        file: StaticString = #file,
+        line: UInt = #line
+    ) throws -> T.Output {
```
이 메서드의 사용용도에 대해 조금 설명해주실 수 있을까요?

derrickkim0109 commented on Tuist/ProjectDescriptionHelpers/SPM.swift:
```diff
@@ -15,6 +15,9 @@ public extension TargetDependency.SPM {
     static let Kingfisher = Self.package(product: "Kingfisher")
     static let FlexLayout = Self.package(product: "FlexLayout")
     static let PinLayout = Self.package(product: "PinLayout")
+	static let KakaoSDKCommon = Self.package(product: "KakaoSDKCommon")
+	static let KakaoSDKAuth = Self.package(product: "KakaoSDKAuth")
+	static let KakaoSDKUser = Self.package(product: "KakaoSDKUser")
```
저희 저번에 정렬 맞추지 않았었나요? 
위에 코멘트 남겼는데 저번이랑 비슷한 상황인 것 같아요.

정렬이 되있는 상황이라면 패스하셔도 됩니다~

inwoodev replied:
수정했습니다~

derrickkim0109 reviewed (commented):
수고하셨습니다~!

궁금한 점에 대해 코멘트 남겼습니다.
머지할 때 컴플릭이 많이 발생할 것 같은데 일요일에 같이 해보시죠~

inwoodev committed:
chore: Tuist에서 tab기반 indentation -> space기반 indentation으로 수정

derrickkim0109 commented on Plugins/DependencyPlugin/ProjectDescriptionHelpers/Modules.swift:
```diff
@@ -11,6 +11,7 @@ import ProjectDescription
 public enum ModulePath {
     case feature(Feature)
     case coordinator(Coordinator)
+    case dependencyInjection(DependencyInjection)
     case domain(Domain)
     case core(Core)
```
구두로 논의 끝에 Dependency Injection은 외부 모듈로 빼는 걸로 결정

inwoodev committed:
fix: 누락된 DomainTests 타겟 추가

inwoodev committed:
fix: Tuist 모듈정리

- DependencyInjection 의존성을 Feature에 추가
- FlexLayout dependency를 Coordinator, Feature, 그리고 App에 추가
  - App 제외 빌드 오류가 발생하는 상황
- allLoad세팅 추가(적용x)
  - Flexlayout을 모든 모듈에 추가해야하는 문제를 해결하기 위해 .allLoad세팅을 Target세팅에 추가하려 하였으나 duplicate symbol 문제 때문에 보류
